src/Features/LanguageServer/Protocol/Handler/Completion/CompletionListCache.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Completion
+{
+    /// <summary>
+    /// Caches completion lists in between calls to CompletionHandler and
+    /// CompletionResolveHandler. Used to avoid unnecessary recomputation.
+    /// </summary>
+    [Export(typeof(CompletionListCache)), Shared]
+    internal class CompletionListCache
+    {
+        /// <summary>
+        /// Maximum number of completion lists allowed in cache. Must be >= 1.
+        /// </summary>
+        private const int MaxCacheSize = 3;
+
+        /// <summary>
+        /// Multiple cache requests or updates may be received concurrently.
+        /// We need this sempahore to ensure that we aren't making concurrent
+        /// modifications to _nextResultId or the _resultIdToCompletionList
+        /// dictionary.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new(1);
+
+        #region protected by _semaphore
+        /// <summary>
+        /// The next resultId available to use.
+        /// </summary>
+        private long _nextResultId;
+
+        /// <summary>
+        /// Maps a resultId to its completion list.
+        /// </summary>
+        private readonly Dictionary<long, CompletionList> _resultIdToCompletionList = new();
+        #endregion
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CompletionListCache()
+        {
+        }
+
+        /// <summary>
+        /// Adds a completion list to the cache. If the cache reaches its maximum size, the oldest completion
+        /// list in the cache is removed.
+        /// </summary>
+        /// <returns>
+        /// The generated resultId associated with the passed in completion list.
+        /// </returns>
+        internal async Task<long> UpdateCacheAsync(CompletionList completionList, CancellationToken cancellationToken)
+        {
+            using (await _semaphore.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                // Getting the generated unique resultId
+                var resultId = _nextResultId++;
+
+                // If cache exceeds maximum size, remove the oldest list in the cache
+                if (_resultIdToCompletionList.Count >= MaxCacheSize)
+                {
+                    var oldestCachedResultId = resultId - MaxCacheSize;
```

Comment:
i don't understand or trust this math :)  so i would just say that you can just remove the value with the lowest ID :)